### PR TITLE
Layers View: Added Selection and Allocation functions to each layer segment.

### PR DESCRIPTION
**Pull request by OrionsGuardian**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Added buttons to the Layer Segments to allow users to Select/De-Select graph elements that are allocated to the corresponding layer.
Also added buttons to Allocate/De-Allocate currently selected graph elements to or from the corresponding layer.

### Alternate Designs

-

### Why Should This Be In Core?

Core functionality expected in a Layers View.

### Benefits

Enhances the functionality of Layers View.

### Possible Drawbacks

-

### Verification Process

Create a graph with multiple nodes and multiple layers.
Open the layer segments but do not activate any layers yet.
Individually select some nodes then click on an Allocate button in a specific layer segment.
Now activate only that layer segment and confirm the only nodes showing are the ones that were just allocated to that layer.
Click on the De-Select button ... confirm all nodes are no longer selected.
Click the select button ... confirm that only the nodes allocated to the layer have been selected.
click the De-Allocate button and all the nodes should now disappear from view, as they are now not part of the active layer.
Click the Allocate button, and the node will re-appear, as they are now once again part of the active layer.

If you now uncheck the Include Hidden button and try to repeat the last 2 steps: De-Allocate and then Allocate
The Allocate will not bring back the selected entries because they are hidden from view.

### Applicable Issues

#2182

### CODE REVIEW

**Review by Delphinus8821 — APPROVED** [review]

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java [hunk 1]
```diff
@@ -132,12 +139,55 @@ private Pane createLayerDetailsPane(final BitMaskQuery query) {
         vxQuery = new QueryInputPane(this, "Vertex Query: ", VERTEX_DESCRIPTION, query.getQueryElementType() == GraphElementType.VERTEX ? query.getQueryString() : StringUtils.EMPTY, true);
         txQuery = new QueryInputPane(this, "Transaction Query: ", TRANSACTION_DESCRIPTION, query.getQueryElementType() == GraphElementType.TRANSACTION ? query.getQueryString() : StringUtils.EMPTY, true);
         descinput = new QueryInputPane(this, "Query Description", QUERY_DESCRIPTION, query.getDescription(), false);
-        final VBox box = new VBox(10, descinput, vxQuery, txQuery);
+        final CheckBox includeHidden = new CheckBox("Include Hidden");
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/components/LayerTitlePane.java:142]: are you able to some newlines between each button block just for readability

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/BitMaskQueryCollection.java [hunk 2]
```diff
@@ -85,6 +86,33 @@ public void setQuery(final String query, final int bitMaskIndex) {
         queries[bitMaskIndex] = new BitMaskQuery(new Query(elementType, query), bitMaskIndex, StringUtils.EMPTY);
     }
 
+    
+    public long getActiveQueriesBitmask(){
+        return activeQueriesBitMask;
+    }
+    
+    public List<Boolean> getVisibilityList() {
+        List<Boolean> results = new ArrayList<>();
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/BitMaskQueryCollection.java:95]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/BitMaskQueryCollection.java [hunk 3]
```diff
@@ -85,6 +86,33 @@ public void setQuery(final String query, final int bitMaskIndex) {
...
+            if (query != null) {
+                results.add(query.isVisible());
+            } else {
+                results.add(false);
+            }
+        }
+        return results;
+    }
+    
+    public void setVisibilities(List<Boolean> revisedVis) {
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/query/BitMaskQueryCollection.java:106]: parameter final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java [hunk 4]
```diff
@@ -111,4 +123,178 @@ public static Button createHelpButton() {
 
         return helpDocumentationButton;
     }
+
+    public static void selectVisibleElements(final boolean selectionSetting){
+        Thread elementSelecter = new Thread(() -> directSelectVisibleElements(selectionSetting));
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java:128]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java [hunk 5]
```diff
@@ -111,4 +123,178 @@ public static Button createHelpButton() {
...
+        elementSelecter.start();
+    }
+    
+    public static void selectLayerElements(final int layerBitMap, final boolean selectionSetting, final boolean includeHidden) {
+        Thread layerSelecter = new Thread(() -> directSelectLayerElements(layerBitMap, selectionSetting, includeHidden));
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java:133]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java [hunk 6]
```diff
@@ -111,4 +123,178 @@ public static Button createHelpButton() {
...
+        layerSelecter.start();
+    }
+
+    public static void allocateElementsForLayer(final int layerBitMap, final boolean allocationSetting, final boolean includeHidden) {
+        Thread layerSelecter = new Thread(() -> directAllocateElementsForLayer(layerBitMap, allocationSetting, includeHidden));
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersUtilities.java:138]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

**Pull request merged**